lib/fs/basicfs_symlink_windows.go

```diff
@@ -0,0 +1,197 @@
+	base := os.TempDir()
+	path := filepath.Join(base, "symlinktest")
+	defer os.Remove(path)
+
+	err := DefaultFilesystem.CreateSymlink(path, base, LinkTargetDirectory)
+	if err != nil {
+		return
+	}
+
+	stat, err := osutil.Lstat(path)
+	if err != nil || stat.Mode()&os.ModeSymlink == 0 {
+		return
+	}
+
+	target, tt, err := DefaultFilesystem.ReadSymlink(path)
+	if err != nil || osutil.NativeFilename(target) != base || tt != LinkTargetDirectory {
+		return
+	}
+	symlinksSupported = true
+}
+
+func DisableSymlinks() {
+	symlinksSupported = false
+}
+
+func (BasicFilesystem) SymlinksSupported() bool {
+	return symlinksSupported
+}
+
+func (BasicFilesystem) ReadSymlink(path string) (string, LinkTargetType, error) {
+	ptr, err := syscall.UTF16PtrFromString(path)
+	if err != nil {
+		return "", LinkTargetUnknown, err
+	}
+	handle, err := syscall.CreateFile(ptr, 0, syscall.FILE_SHARE_READ|syscall.FILE_SHARE_WRITE|syscall.FILE_SHARE_DELETE, nil, syscall.OPEN_EXISTING, syscall.FILE_FLAG_BACKUP_SEMANTICS|win32FileFlagOpenReparsePoint, 0)
+	if err != nil || handle == syscall.InvalidHandle {
+		return "", LinkTargetUnknown, err
+	}
+	defer syscall.Close(handle)
+	var ret uint16
+	var data reparseData
+
+	r1, _, err := syscall.Syscall9(procDeviceIoControl.Addr(), 8, uintptr(handle), win32FsctlGetReparsePoint, 0, 0, uintptr(unsafe.Pointer(&data)), unsafe.Sizeof(data), uintptr(unsafe.Pointer(&ret)), 0, 0)
+	if r1 == 0 {
+		return "", LinkTargetUnknown, err
+	}
+
+	tt := LinkTargetUnknown
+	if attr, err := syscall.GetFileAttributes(ptr); err == nil {
+		if attr&syscall.FILE_ATTRIBUTE_DIRECTORY != 0 {
+			tt = LinkTargetDirectory
+		} else {
+			tt = LinkTargetFile
+		}
+	}
+
+	return osutil.NormalizedFilename(data.printName()), tt, nil
+}
+
+func (BasicFilesystem) CreateSymlink(path, target string, tt LinkTargetType) error {
+	srcp, err := syscall.UTF16PtrFromString(path)
+	if err != nil {
+		return err
+	}
+
+	trgp, err := syscall.UTF16PtrFromString(osutil.NativeFilename(target))
+	if err != nil {
+		return err
+	}
+
+	// Sadly for Windows we need to specify the type of the symlink,
+	// whether it's a directory symlink or a file symlink.
+	// If the flags doesn't reveal the target type, try to evaluate it
+	// ourselves, and worst case default to the symlink pointing to a file.
+	mode := 0
+	if tt == LinkTargetUnknown {
+		path := target
+		if !filepath.IsAbs(target) {
+			path = filepath.Join(filepath.Dir(path), target)
+		}
+
+		stat, err := os.Stat(path)
+		if err == nil && stat.IsDir() {
+			mode = win32SymbolicLinkFlagDirectory
+		}
+	} else if tt == LinkTargetDirectory {
+		mode = win32SymbolicLinkFlagDirectory
+	}
+
+	r0, _, err := syscall.Syscall(procCreateSymbolicLink.Addr(), 3, uintptr(unsafe.Pointer(srcp)), uintptr(unsafe.Pointer(trgp)), uintptr(mode))
+	if r0 == 1 {
+		return nil
+	}
+	return err
+}
+
+func (fs BasicFilesystem) ChangeSymlinkType(path string, tt LinkTargetType) error {
+	LinkTarget, exTt, err := fs.ReadSymlink(path)
```

Comment:
the uppercase here is weird, also this exTt is weird.

target, existingTargetType, err ?